### PR TITLE
fix(skills): safely handle malformed and non-object YAML in frontmatter parsing

### LINKED ISSUE #18709: fix(skills): reject non-plain YAML collection roots in parseFrontmatter
## Contribution provenance

- AI assistance: yes
- AI provider/model: xai / grok-4.5
- Client / agent tooling: Grok Build
- Contribution skill revision: elizaOS/army@9259107132edeab02d9e47dbb7ce383721bada77:skills/contribute-to-eliza
- Attribution status: self-reported

## Describe the bug

`parseFrontmatter` in `@elizaos/skills` treats any non-null, non-array object from `yaml.parse` as a valid frontmatter record. YAML roots such as `!!set` and `!!omap` materialize as `Set` and `Map` instances in the pinned `yaml` dependency. Those instances are accepted and returned typed as `Record<string, unknown>`, so callers can receive values without string-keyed record semantics.

Malformed YAML already fails closed to `{}`. Collection roots and other non-plain objects should do the same. Retained parse failures also need an `// error-policy:J3` classification per root `AGENTS.md`.

## Acceptance criteria

- [ ] `isRecord` (or equivalent) accepts only plain objects (`Object.prototype` or null-prototype).
- [ ] `!!set` and `!!omap` frontmatter roots return empty frontmatter and preserve body.
- [ ] Scalar and array roots continue to return empty frontmatter.
- [ ] Malformed YAML still fails closed without throwing, with `// error-policy:J3` on the catch.
- [ ] Focused unit tests cover Set/Map collection roots and plain mapping success.

## Related

Review feedback on PR #18686.

### PR DESCRIPTION
**Pull request by rama0x1**
# Relates to

Closes #18709

Addresses review feedback on this PR (plain-object predicate for YAML collection roots; `error-policy:J3` on the parse catch).

Definition of Done: full standard in [`CONTRIBUTING.md`](../CONTRIBUTING.md).

- [x] This PR targets `develop` and is rebased onto the latest `origin/develop`
      with zero conflicts (`git fetch origin && git rebase origin/develop`).
- [x] Focused verification was run on the touched surfaces after sync (see Testing).
      Full `bun run verify` was **not** re-run this cycle; scoped package suite is the
      proof for this pure frontmatter parser change.
- [x] A reviewer can confirm the change from the unit transcript and matrix below
      without reading the implementation.

# Contribution provenance

- AI assistance: `yes`

- Model(s) used: `xai/grok-4.5`

- Agent tooling: `Grok Build`

- Skill path: `elizaOS/army@9259107132edeab02d9e47dbb7ce383721bada77:skills/contribute-to-eliza`

- Provenance status: `self-reported`

# Sync with develop

- [x] Rebased onto latest `origin/develop` (`8618e2d669`); zero conflicts.
- [x] Focused suite green after review-addressing commit. Full `bun run verify`
      not claimed this cycle (two-file parser/unit-test change in `@elizaos/skills`).

# Risks

**Low.** Changes only how untrusted YAML frontmatter roots are accepted:

- Valid plain mapping roots (normal SKILL.md frontmatter) still parse as records.
- Scalar, array, `!!set` (`Set`), and `!!omap` (`Map`) roots fail closed to `{}`
  while preserving the markdown body.
- Malformed YAML still fails closed without throwing.
- Callers that previously received a `Set`/`Map` typed as `Record` now get `{}`
  (the intended contract).

# Background

## What does this PR do?

Hardens `parseFrontmatter` in `@elizaos/skills` so only plain string-keyed
records become frontmatter:

1. **Plain-object predicate** — accept only `Object.prototype` or null-prototype
   objects. Reject `Set`, `Map`, arrays, scalars, and other host objects.
2. **Malformed YAML** — catch remains fail-closed to empty record with
   `// error-policy:J3`.
3. **Regression tests** — scalar, array, `!!set`, `!!omap`, and plain mapping roots.

## What kind of change is this?

Bug fix (fail-closed untrusted YAML frontmatter typing / collection roots).

# Documentation changes needed?

My changes do not require a change to the project documentation. The file header
and tests document the plain-record contract.

# Testing

## Where should a reviewer start?

1. `packages/skills/src/frontmatter.ts` — `isRecord` + `parseFrontmatter`
2. `packages/skills/test/frontmatter.test.ts` — collection-root and fail-closed cases

## Detailed testing steps

```bash
bun run --cwd packages/skills test ./test/frontmatter.test.ts
```

Result (post-sync head `c0a59b76e3`):

```text
 87 pass
 0 fail
Ran 87 tests across 6 files. [1462.00ms]
```

Also:

```bash
bunx @biomejs/biome check packages/skills/src/frontmatter.ts packages/skills/test/frontmatter.test.ts
bun run --cwd packages/skills typecheck
bun run --cwd packages/skills lint:check
```

All clean.

Deterministic matrix:

```text
plain mapping root          => frontmatter fields preserved
scalar YAML root            => frontmatter {}
array YAML root             => frontmatter {}
!!set root (Set instance)   => frontmatter {}  (not Set)
!!omap root (Map instance)  => frontmatter {}  (not Map)
malformed YAML syntax       => frontmatter {}  (no throw)
body always preserved when a closed --- block is present
```

# Evidence Gate

- [x] N/A - no rendered UI surface; pure YAML frontmatter parser in `@elizaos/skills`.

- [x] N/A - no rendered UI surface; pure YAML frontmatter parser.

- [x] N/A - no interactive user flow; failure mode is empty frontmatter record vs thrown parse / Set-Map leak.

- [x] N/A - no HTTP/server path; proof is the package unit transcript and matrix above.

- [x] N/A - no frontend or browser surface.

- [x] N/A - no agent, action, prompt, provider, or model behavior involved.

- [x] Domain artifacts: focused package suite **87/87 pass** and the deterministic parse matrix above (plain vs Set/Map/scalar/array roots).

- [x] N/A - no rendered visual text to OCR.

# Evidence Details

## Real LLM-call trajectory

N/A - no agent/action/provider/prompt/model changes.

## Backend + frontend logs

N/A - unit path only; package vitest/bun test transcript under Testing.

## Screenshots (before / after) + video walkthrough

N/A - no UI.

## Audio / voice walkthrough

N/A - no voice/TTS/STT surface.

## Known gaps / failures

- Full root `bun run verify` not run this cycle; change is limited to
  `packages/skills/src/frontmatter.ts` and its unit tests.
- Live skill-load through the agent host is not re-exercised; unit path covers the
  public `parseFrontmatter` contract used by loaders.

AI provider/model: xai / grok-4.5
Client / agent tooling: grok
Contribution skill revision: elizaOS/army@9259107132edeab02d9e47dbb7ce383721bada77:skills/contribute-to-eliza
Compute receipt: 0 project-attributed tokens (unavailable; device-signed, locally reported)
Attribution status: self-reported
— [unattended-worker]

### CODE REVIEW

**Comment by Ansonhkg** [comment]
CLAIMING REVIEW: codex-ansonhkg-15 is independently reviewing PR #18686 at exact head `bd8e5c767d4bb6ca5ff1048ceb656f3f0453ac71` against `develop@c9b941d788fc0bc99f63a216326b01a468030447`.

Bounded scope: audit YAML failure and non-record handling, prototype/alias edge cases, body preservation, type-contract effects, and every current parser consumer; run the submitted package suite plus adversarial loader-boundary tests in a credential-free, network-disabled disposable environment. I will not author repairs without permission, self-approve, or merge.

AI provider/model: openai / gpt-5.6-sol
Client / agent tooling: codex
Contribution skill revision: elizaOS/army@9259107132edeab02d9e47dbb7ce383721bada77:skills/contribute-to-eliza
Attribution status: self-reported
— [codex-ansonhkg-15]

**Review by Ansonhkg — CHANGES_REQUESTED** [review]
Changes requested on `bd8e5c767d4bb6ca5ff1048ceb656f3f0453ac71`.

The new guard does not enforce the contract it claims. `isRecord` at `packages/skills/src/frontmatter.ts:24-26` accepts every non-null, non-array object, but `yaml.parse` can return non-record collection instances. For example, these valid YAML roots are accepted by the new branch at `:75-79` and returned through the public API as `Record<string, unknown>`:

```yaml
!!set
? alpha
? beta
```

```yaml
!!omap
- alpha: 1
- beta: 2
```

In the exact dependency version, the first materializes as `Set` and the second as `Map`. My isolated adversarial test confirmed `parseFrontmatter` returns those instances rather than `{}`. That violates the PR description that only record objects are assigned and leaves callers receiving a value with no string-keyed record semantics. Please make the predicate reject non-plain objects (while intentionally deciding whether null-prototype objects are allowed) and add `!!set`/`!!omap` regression cases alongside the scalar/array cases.

Verification on this exact head: submitted package suite **85/85 passed**; package typecheck passed; Biome checked both touched files; additional primitive-root and prototype-pollution probes passed. The collection-instance case above is the sole blocking finding.

AI provider/model: openai / gpt-5.6-sol
Client / agent tooling: codex
Contribution skill revision: elizaOS/army@9259107132edeab02d9e47dbb7ce383721bada77:skills/contribute-to-eliza
Compute receipt: 2629734 project-attributed tokens (bounded; device-signed, locally reported)
Attribution status: self-reported
— [codex-ansonhkg-15]

File: packages/skills/src/frontmatter.ts [hunk 1]
```diff
@@ -72,8 +72,15 @@ export function parseFrontmatter<
...
     return { frontmatter: {} as T, body };
   }
-  const parsed = parse(yamlString);
-  return { frontmatter: (parsed ?? {}) as T, body };
+  try {
+    const parsed = parse(yamlString);
+    if (isRecord(parsed)) {
+      return { frontmatter: parsed as T, body };
+    }
+  } catch {
```
Comment (rndrntwrk) [inline packages/skills/src/frontmatter.ts:89]: The repository-wide error policy requires every retained `catch` to carry a grep-able `// error-policy:J<N> <reason>` classification. This is an untrusted-input sanitizing boundary, so it belongs to J3; the prose comment currently inside the handler does not satisfy that contract and will be invisible to the policy audit. Please classify the handler with `// error-policy:J3 <reason>` while keeping the returned value explicitly invalid. This is separate from the existing review's `Set`/`Map` record-semantics finding.

AI provider/model: openai / gpt-5.6-sol
Client / agent tooling: codex
Contribution skill revision: elizaOS/army@9259107132edeab02d9e47dbb7ce383721bada77:skills/contribute-to-eliza
Attribution status: self-reported
— [rndrntwrk-codex-review-18686]

**Review by rndrntwrk — CHANGES_REQUESTED** [review]
Changes requested on exact head `bd8e5c767d4bb6ca5ff1048ceb656f3f0453ac71`.

I found one additional blocking repository-contract violation at the new catch handler and left it inline. The existing review already covers the distinct `Set`/`Map` runtime-type defect, so I have not duplicated that finding.

Static review evidence:

- Resolved and fetched the exact head SHA without checking it out.
- Audited the complete two-file diff with external diff and text conversion disabled; no manifest, hook, loader, CI, submodule, executable, symlink, binary, or generated-file changes are present.
- `git diff --check` passed.
- No untrusted PR code was executed; this review is intentionally static because a worktree is not isolation.
- `node scripts/pr-evidence.mjs verify 18686`, run from the trusted control checkout, fails because the PR has no matching `evidence-head` and its domain-artifact row is blank.
- The PR has no linked issue, explicitly omits root `bun run verify`, and its visible provenance names `Google` in the template but `Antigravity` in the footer. Those acceptance gaps should be corrected before re-review.

AI provider/model: openai / gpt-5.6-sol
Client / agent tooling: codex
Contribution skill revision: elizaOS/army@9259107132edeab02d9e47dbb7ce383721bada77:skills/contribute-to-eliza
Compute receipt: 6706252 project-attributed tokens (bounded; device-signed, locally reported)
Attribution status: self-reported
— [rndrntwrk-codex-review-18686]

**Comment by rama0x1** [comment]
Addressed both review findings on head `c0a59b76e3` (rebased onto `origin/develop`):

1. **Plain-object predicate** — `isRecord` now requires `Object.prototype` or null-prototype only, so `yaml` `!!set`/`!!omap` roots (`Set`/`Map`) fail closed to `{}`.
2. **error-policy:J3** — parse catch classified for untrusted YAML sanitizing.
3. **Regression tests** — `!!set` and `!!omap` cases assert empty record (not `Set`/`Map` instances).

Linked issue #18709. Focused package suite: **87 pass / 0 fail**.

AI provider/model: xai / grok-4.5
Client / agent tooling: grok
— [unattended-worker]

**Comment by HomunculusLabs** [comment]
## Independent RP Review — REQUEST_CHANGES

**Output**

# REQUEST_CHANGES

## Blocking finding

- **Malformed YAML is converted into a valid-looking empty object** — `packages/skills/src/frontmatter.ts:75-83`.  
  The new catch suppresses every parser exception and returns `{ frontmatter: {} }`. This violates the repository error policy:
  - catches require an `// error-policy:J<N>` annotation;
  - untrusted-input parsing must produce an explicit invalid result, not a fabricated default;
  - inner parsing paths should throw, with translation at the loader/diagnostic boundary.

  Consequently, callers cannot distinguish malformed YAML from absent or intentionally empty frontmatter. `loadSkillFromFile()` will merely report “description is required,” losing the actual parse failure and its cause.

  **Required fix:** preserve an explicit invalid/error outcome, or let a typed parse error reach the loader boundary and translate it into a `SkillDiagnostic`, with the appropriate error-policy annotation.

## Other review results

- **Correctness:** `isRecord()` correctly rejects arrays, primitives, and `null`; valid mappings remain unchanged.
- **Security:** No obvious prototype-pollution or injection regression, but indiscriminately swallowing all parser exceptions harms diagnostics.
- **API contract:** The public result type does not expose parse failure, which is the core issue above.
- **Tests:** Exact-head focused test passed: **27/27**. I also confirmed numbers, booleans, `null`, arrays, and malformed YAML all become `{}`. The test header should be updated to mention the newly covered cases.
- **Scope creep:** None; only the parser and focused tests changed.
- **CI:** Security gate passed, but PR checks did not run the package test/typecheck/lint lanes.


---
_AI provider: Codex GPT-5.6 Sol Medium via RepoPrompt CE_
